backend-tests/tests/test_deployments.py

```diff
@@ -502,3 +583,315 @@ def compare_phases_json(expected, response):
+        ### Static helper function ###
+        # Setup Deployment APIs
+        api_mgmt_deploy = ApiClient(deployments.URL_MGMT)
+
+        devices_updated = 0
+        num_phases = len(deployment["phases"])
+        batch_sizes = [int((deployment["phases"][i]["batch_size"] / 100.0) \
+                             * len(devices)) for i in range(num_phases-1)]
+        # Final batch_size might not be specified
+        batch_sizes.append(len(devices) - sum(batch_sizes))
+
+        # POST deployment
+        resp = api_mgmt_deploy.with_auth(user_token).call(
+            'POST',
+            deployments.URL_DEPLOYMENTS,
+            body=deployment
+        )
+        assert resp.status_code == 201
+
+        # Store the location from the GET /deployments/{id} request
+        deployment_id = os.path.basename(resp.headers['Location'])
+
+        for i in range(num_phases):
+            if i == 0 and "start_ts" not in deployment["phases"][i]:
+                # First phase don't need to specify `start_ts`
+                pass
+            elif "start_ts" in deployment["phases"][i]:
+                # Sleep until next phase starts
+                start_ts = datetime.strptime(
+                    deployment["phases"][i]["start_ts"],
+                    "%Y-%m-%dT%H:%M:%SZ"
+                )
+                now = datetime.utcnow()
+                wait_time = start_ts - now
+
+                # While phase in progress
+                # NOTE: add a half a second buffer time, as a just-in-time
+                #       request will break the remainder of the test
+                while now < (start_ts - timedelta(milliseconds=500)):
+                    # Spam update requests from random non-updated devices
+                    dev = random.choice(devices[devices_updated:])
+                    try_update(dev, expected_status_code=204)
+                    now = datetime.utcnow()
+                # Sleep the last 500ms to let the next phase start
+                time.sleep(0.5)
+            else:
+                raise ValueError("Invalid phased deployment request, "
+                                 "missing `start_ts` for phase %d" % i)
+
+            ### Test for all devices in the deployment ###
+            if devices_updated > 0:
+                # Allready updated
+                for device in devices[:devices_updated]:
+                    try_update(
+                        device,
+                        artifact_name=deployment["artifact_name"],
+                        expected_status_code=204
+                    )
+
+            # Check phase count has not been incremented by the above requests
+            resp = api_mgmt_deploy.with_auth(
+                user_token).call(
+                    'GET',
+                    deployments.URL_DEPLOYMENT.format(id=deployment_id)
+                )
+            phase = resp.json()["phases"][i]
+            assert phase["device_count"] == 0
+
+            if batch_sizes[i] > 0:
```

Comment:
nitpick - won't this always be true? we fill up batch size with known batch size, and also - with an extra one if the last phase doesn't have it.

Reply:
Yes it is, but it is legal to use `batch_size` of zero? Then perhaps we should include a test case for this special case aswell?